vminitd/Sources/vminitd/VsockProxy.swift

```diff
@@ -191,57 +235,79 @@ extension VsockProxy {
                 }
 
                 try! ProcessSupervisor.default.poller.add(clientFile.fileDescriptor, mask: EPOLLIN | EPOLLOUT) { mask in
-                    if mask.readyToRead {
-                        do {
-                            let (_, _, action) = try OSFile.splice(from: &clientFile, to: &serverFile)
-                            if action == .eof || action == .brokenPipe {
-                                return cleanup()
-                            }
-                        } catch {
-                            return cleanup()
-                        }
+                    if mask.readyToRead && !eofFromClient {
+                        let (fromEof, toEof) = Self.transferData(
+                            fromFile: &clientFile,
+                            toFile: &serverFile,
+                            description: "readyToRead:toServer",
+                            log: self.log
+                        )
+                        eofFromClient = eofFromClient || fromEof
+                        eofFromServer = eofFromServer || toEof
                     }
 
-                    if mask.readyToWrite {
-                        do {
-                            let (_, _, action) = try OSFile.splice(from: &serverFile, to: &clientFile)
-                            if action == .eof || action == .brokenPipe {
-                                return cleanup()
-                            }
-                        } catch {
-                            return cleanup()
-                        }
+                    if mask.readyToWrite && !eofFromServer {
+                        let (fromEof, toEof) = Self.transferData(
+                            fromFile: &serverFile,
+                            toFile: &clientFile,
+                            description: "readyToWrite:toClient",
+                            log: self.log
+                        )
+                        eofFromClient = eofFromClient || toEof
+                        eofFromServer = eofFromServer || fromEof
                     }
 
                     if mask.isHangup {
+                        eofFromClient = true
+                        eofFromServer = true
+                    } else if mask.isRhangup && !eofFromClient {
+                        // half close, shut down client to server transfer
+                        // we should see no more EPOLLIN events on the client fd
+                        // and no more EPOLLOUT events on the server fd
+                        eofFromClient = true
+                        shutdown(serverFile.fileDescriptor, SHUT_WR)
```

Comment:
Check if failed and log something (as we can't really bubble anything up from here).